enterprise/frontend/src/metabase-enterprise/content_translation/utils.ts

```diff
@@ -0,0 +1,175 @@
+import * as I from "icepick";
+import { useMemo } from "react";
+import _ from "underscore";
+
+import type { ContentTranslationFunction } from "metabase/i18n/types";
+import type { HoveredObject } from "metabase/visualizations/types";
+import type {
+  DatasetColumn,
+  DictionaryArray,
+  Series,
+} from "metabase-types/api";
+
+import { useTranslateContent } from "./use-translate-content";
+
+export type TranslateContentStringFunction = <
+  MsgidType = string | null | undefined,
+>(
+  dictionary: DictionaryArray | undefined,
+  locale: string | undefined,
+  /** This argument will be translated only if it is a string. If it is not a
+   * string, it will be returned untranslated. */
+  msgid: MsgidType,
+) => string | MsgidType;
+
+/** Translate a user-generated string
+ *
+ * Terminology: A "msgid" is a 'raw', untranslated string. A "msgstr" is a
+ * translation of a msgid.
+ * */
+export const translateContentString: TranslateContentStringFunction = (
+  dictionary,
+  locale,
+  msgid,
+) => {
+  if (!locale) {
+    return msgid;
+  }
+
+  if (typeof msgid !== "string") {
+    return msgid;
+  }
+
+  if (!msgid.trim()) {
+    return msgid;
+  }
+
+  const msgstr = dictionary?.find(
+    (row) => row.locale === locale && row.msgid === msgid,
+  )?.msgstr;
```

Comment:
I don't think we can do an O(n * m) lookup per render (n = number of items to translate, m = size of your dictionary) for views with large datasets. Something like `type TCDictionary<locale, Record<msgid, translation>>` is going to serve us better I think O(n).

And as always with perf... measure! Ideally using as real world of a dictionary as possible. With a tiny enough dictionary I'd expect that this code as is would beat out my suggestion.